File: src/jsonSchemaInspector.ts
import {ExtendedJSONSchema7, log_error, log_warning, SchemaKeyProperties} from "singer-node"
import {JSONSchema7, JSONSchema7Definition, JSONSchema7TypeName} from "json-schema"
import {asArray} from "./utils"
import SchemaTranslator, {ValueTranslator} from "./SchemaTranslator"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const get = require("lodash.get")

// eslint-disable-next-line @typescript-eslint/no-var-requires
const sha1 = require('sha1')

// To use some non-standard property
export interface IExtendedJSONSchema7 extends ExtendedJSONSchema7 {
  decimals?: number;
  precision?: number;
  lowCardinality?: boolean
}

// Least likely to appear in column name
export const nestedSubObjectSeparator = "$%€£"

export class JsonSchemaInspectorContext {

  constructor(
    public readonly alias: string,
    public readonly schema: IExtendedJSONSchema7,
    public readonly keyProperties: string[], // For current level. Only root has if all_key_properties isn't defined
    public readonly subtableSeparator = "__",
    public readonly parentCtx?: JsonSchemaInspectorContext,
    public readonly level: number = 0,
    public readonly tableName = JsonSchemaInspectorContext.defaultTableName(alias, subtableSeparator, parentCtx),
    public readonly cleaningColumn?: string,
    // Optional config to know all key properties at this level and lower. Used to compute _parent_... fields
    public readonly allKeyProperties: SchemaKeyProperties = {props: [], children: {}},
  ) {
  }

  static defaultTableName(alias: string, subtableSeparator: string, parentCtx?: JsonSchemaInspectorContext): string {
    return `${parentCtx ? `${parentCtx.tableName}${subtableSeparator}` : ""}${alias}`
  }

  public isTypeObject() {
    return this.schema.type?.includes("object") ?? false
  }

  public isRoot() {
    return this.parentCtx === undefined
  }

  public getRootContext(): JsonSchemaInspectorContext {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.isRoot() ? this : this.parentCtx!.getRootContext()
  }
}

// Can be any in case of arrays without PK
export type ValueExtractor = (data: Record<string, any> | any) => any

export interface ISimpleColumnType {
  chType?: string;
  valueTranslator?: ValueTranslator;
  typeFormat?: string;
  nullable: boolean,
  lowCardinality: boolean,
  nestedArray?: boolean
}

export interface IColumnMapping {
  prop?: string;
  valueExtractor: ValueExtractor;
  sqlIdentifier: string;
}

export type ColumnMap = IColumnMapping & ISimpleColumnType;

// eslint-disable-next-line no-shadow
export enum PKType {
  ROOT = "ROOT",
  PARENT = "PARENT",
  CURRENT = "CURRENT",
  LEVEL = "LEVEL",
}

export interface IPKMapping {
  prop: string;
  valueExtractor: ValueExtractor;
  sqlIdentifier: string;
  pkType: PKType
}

export type PkMap = IPKMapping & ISimpleColumnType;

export interface ISourceMeta {
  prop: string
  children: ISourceMeta[];
  pkMappings: PkMap[];
  simpleColumnMappings: ColumnMap[];
  sqlTableName: string;
  cleaningColumn?: string;
}

export const formatLevelIndexColumn = (level: number) => `_level_${level}_index`
export const formatRootPKColumn = (prop: string) => `_root_${prop}`
export const formatParentPKColumn = (prop: string) => `_parent_${prop}`

const buildMetaPkProp = (prop: string, ctx: JsonSchemaInspectorContext, pkType: PKType, fieldFormatter?: (v: string) => string): PkMap => ({
  prop,
  valueExtractor: buildValueExtractor(prop),
  sqlIdentifier: escapeIdentifier(fieldFormatter?.(prop) ?? prop, ctx.subtableSeparator),
  ...getSimpleColumnType(ctx, prop),
  nullable: false,
  lowCardinality: false,
  nestedArray: false,
  pkType,
})

const buildValueExtractor = (prop: string | undefined): ValueExtractor => {
  if (prop) {
    const propParts = prop.split(nestedSubObjectSeparator)
    if (propParts.length == 1) {
      const uniqPart = propParts[0]
      return (data) => data[uniqPart]
    } else {
      return (data) => get(data, propParts)
    }
  } else {
    return (data) => data
  }
}

const buildMetaPkProps = (ctx: JsonSchemaInspectorContext): PkMap[] => ([] as PkMap[])
  // Append '_root_X'
  .concat(ctx.isRoot() ? [] : ctx.getRootContext().keyProperties.map((prop => buildMetaPkProp(prop, ctx.getRootContext(), PKType.ROOT, formatRootPKColumn))))
  // Append '_parent_X' if parent has 'all_key_properties' filled with props
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  .concat((!(ctx.parentCtx?.allKeyProperties?.props.length === 0) && ctx.parentCtx?.keyProperties.map((prop => buildMetaPkProp(prop, ctx.parentCtx!, PKType.PARENT, formatParentPKColumn)))) || [])
  // Append 'X' if defined
  .concat(ctx.keyProperties.map((prop => buildMetaPkProp(prop, ctx, PKType.CURRENT))))
  // Append 'level_N_index' columns
  .concat(Array.from(Array(ctx.level).keys()).map((value) => {
    const prop = formatLevelIndexColumn(value)
    return {
      prop,
      sqlIdentifier: escapeIdentifier(prop, ctx.subtableSeparator),
      chType: "Int32",
      nullable: false,
      lowCardinality: false,
      nestedArray: false,
      pkType: PKType.LEVEL,
    } as PkMap
  }))

// transform a schema to a data structure with metadata for Clickhouse (types, primary keys, nullable, ...)
export const buildMeta = (ctx: JsonSchemaInspectorContext): ISourceMeta => ({
  prop: ctx.alias,
  sqlTableName: escapeIdentifier(ctx.tableName, ctx.subtableSeparator),
  pkMappings: buildMetaPkProps(ctx),
  cleaningColumn: ctx.cleaningColumn,
  ...buildMetaProps(ctx),
})

function makeNullable(type?: JSONSchema7TypeName | JSONSchema7TypeName[]): JSONSchema7TypeName[] {
  if (!type) {
    return []
  }
  return asArray(type)
    .concat(!type.includes("null") ? ["null"] : [])
}

// flatten 1..1 relation properties into the current level
function flattenNestedObject(propDef: IExtendedJSONSchema7, key: string, ctx: JsonSchemaInspectorContext) {
  const nullable = getNullable(propDef)
  const nestedSchema: IExtendedJSONSchema7 = Object.entries(propDef.properties ?? {})
    .reduce((acc, [nestedKey, nestedPropDef]) => {
    if (typeof nestedPropDef === "boolean") {
      throw new Error("unhandled boolean propdef")
    }
    return {
      ...acc,
      properties: {
        ...acc.properties,
        [`${key}${nestedSubObjectSeparator}${nestedKey}`]: {
          ...nestedPropDef,
          type: nullable ? makeNullable(nestedPropDef.type) : nestedPropDef.type, // if parent is nullable, all children should also be
        },
      },
    }
  }, {type: "object", properties: {}})


  return buildMetaProps(new JsonSchemaInspectorContext(
    ctx.alias,
    nestedSchema,
    [],
    ctx.subtableSeparator,
    ctx,
    ctx.level,
    ctx.tableName,
  ))
}

const createSubTable = (propDef: IExtendedJSONSchema7, key: string, ctx: JsonSchemaInspectorContext): ISourceMeta => buildMeta(new JsonSchemaInspectorContext(
  key,
  (propDef.items || {type: "string"}) as IExtendedJSONSchema7,
  ctx.allKeyProperties.children[key]?.props ?? [],
  ctx.subtableSeparator,
  ctx,
  ctx.level + 1,
  undefined,
  undefined,
  ctx.allKeyProperties.children[key],
))

type MetaProps = { children: ISourceMeta[], simpleColumnMappings: ColumnMap[] }

function buildMetaProps(ctx: JsonSchemaInspectorContext): MetaProps {
  if (ctx.isTypeObject()) {
    return Object.entries(ctx.schema.properties ?? {})
      .filter(([key]) => !ctx.keyProperties.includes(key)) // Exclude values already handled in PK
      .reduce((acc: MetaProps, [key, propDef]) => {
        if (typeof propDef === "boolean") {
          throwError(ctx, "propDef as boolean not supported")
          return acc// needed as ts doesn't understand throwError will always end up throwing
        }
        // JSON Spec supports multiple types. We accept this format but only handle one at a time
        const propDefTypes = asArray(propDef.type)

        if (propDefTypes.includes("object")) {
          const {simpleColumnMappings: nestedSimpleColumnMappings, children: nestedChildren} = flattenNestedObject(propDef, key, ctx)
          return {
            ...acc,
            simpleColumnMappings: acc.simpleColumnMappings.concat(nestedSimpleColumnMappings),
            children: acc.children.concat(nestedChildren),
          }
        } else if (propDefTypes.includes("array") && propDef.format !== "nested") {
          if (ctx.getRootContext().keyProperties.length === 0 && ctx.getRootContext().allKeyProperties.props.length === 0) {
            throwError(ctx, `${key} refused: array child with no root key properties`)
          }
          return {
            ...acc,
            children: [...acc.children, createSubTable(propDef, key, ctx)],
          }
        } else {
          const colType = getSimpleColumnType(ctx, key)

          // Column is a scalar value
          if (colType) {
            return {
              ...acc,
              simpleColumnMappings: [...acc.simpleColumnMappings, {
                prop: key,
                valueExtractor: buildValueExtractor(key),
                sqlIdentifier: escapeIdentifier(key, ctx.subtableSeparator),
                ...colType,
              }],
            }
          } else {
            log_warning(`'${ctx.alias}': '${key}': could not be registered (type '${propDef.type}' unrecognized)`)
            return acc
          }
        }
      }, {simpleColumnMappings: [] as ColumnMap[], children: [] as ISourceMeta[]})
  } else {
    if (!ctx.schema.type) {
      return {
        simpleColumnMappings: [],
        children: [],
      }
    }
    return {
      simpleColumnMappings: [{
        valueExtractor: buildValueExtractor(undefined),
        sqlIdentifier: escapeIdentifier("value", ctx.subtableSeparator),
        ...getSimpleColumnType(ctx, undefined),
        nullable: getNullable(ctx.schema),
        lowCardinality: false,
        nestedArray: false,
      }],
      children: [],
    }
  }
}

// To sanitize "types" keys in JSON Schemas
function excludeNullFromArray(array?: JSONSchema7TypeName | JSONSchema7TypeName[]) {
  return asArray(array).filter((type) => type !== "null")
}

function getNullable(propDef: JSONSchema7Definition) {
  if (typeof propDef === "boolean") {
    throw new Error(`boolean propDef not handled`)
  }
  return asArray(propDef.type).includes("null") ?? false
}

const getLowCardinality = (propDef: IExtendedJSONSchema7) => propDef.lowCardinality !== null && propDef.lowCardinality === true

function getSimpleColumnType(ctx: JsonSchemaInspectorContext, key?: string): ISimpleColumnType | undefined {
  let propDef: JSONSchema7 | boolean | undefined = key ? ctx.schema.properties?.[key] : ctx.schema
  let nestedArray = false
  if (!propDef || typeof propDef === "boolean") {
    throwError(ctx, `Key '${key}' does not match any usable prop in schema props '${ctx.schema.properties}'`)
    return
  }

  if (propDef.format === "nested" && propDef.type === "array") {
    propDef = propDef.items as JSONSchema7
    nestedArray = true
  }
  const type = excludeNullFromArray(propDef.type)[0]
  const chType = getSimpleColumnSqlType(ctx, propDef, key)

  return chType ? {
    valueTranslator: SchemaTranslator.buildTranslator(type),
    typeFormat: propDef.format,
    chType,
    nullable: getNullable(propDef),
    lowCardinality: getLowCardinality(propDef),
    nestedArray,
  } : undefined
}

// From a JSON Schema type, return Clickhouse type
export function getSimpleColumnSqlType(ctx: JsonSchemaInspectorContext, propDef: IExtendedJSONSchema7, key?: string): string | undefined {
  const type = excludeNullFromArray(propDef.type)[0]
  const format = propDef.format
  if (type === "string") {
    if (format === "date" || format === "x-excel-date") {
      return "Date"
    } else if (format === "date-time") {
      return "DateTime"
    } else if (format === "date-time64") {
      return "DateTime64"
    } else if (format === "uuid") {
      return "UUID"
    } else {
      return "String"
    }
  } else if (type === "integer") {
    if (!format) {
      return "Int64"
    } else if (format === "int128") {
      return "Int128"
    } else if (format === "int64") {
      return "Int64"
    } else if (format === "int32") {
      return "Int32"
    } else if (format === "int16") {
      return "Int16"
    } else if (format === "int8") {
      return "Int8"
    } else {
      throwError(ctx, `${key}: unsupported integer format [${format}]`)
    }
  } else if (type === "number") {
    if (!format) {
      return `Decimal(${propDef.precision || 16}, ${propDef.decimals || 2})`
    } else if (format === "float64") {
      return "Float64"
    } else if (format === "float32") {
      return "Float32"
    } else {
      throwError(ctx, `${key}: unsupported number format [${format}]`)
    }
  } else if (type === "boolean") {
    if (!format) {
      return "UInt8" // https://clickhouse.tech/docs/en/sql-reference/data-types/boolean/
    } else {
      throwError(ctx, `${key}: unsupported number format [${format}]`)
    }
  } else {
    return undefined
  }
}

// ensure that id is not longer than 64 chars and enclose it within backquotes
export function escapeIdentifier(id: string, subtableSeparator = "__"): string {
  // @ts-ignore for replaceAll
  id = id.replaceAll(nestedSubObjectSeparator, subtableSeparator)
  if (id.length > 64) {
    const uid = sha1(id).substring(0, 10)
    id = id.substring(0, 64 - uid.length - 27) + uid + id.substring(id.length - 27)
  }
  return `\`${id}\``
}

// throws error with additional context
function throwError(ctx: JsonSchemaInspectorContext, msg: string, childAlias?: string): void {
  const alias = `${ctx.alias}${childAlias ? (`.${childAlias}`) : ""}`
  if (ctx.parentCtx) {
    throwError(ctx.parentCtx, msg, alias)
  } else {
    log_error(`${alias}: ${msg}`)
    throw new Error(`${alias}: ${msg}`)
  }
}

File: src/main/kotlin/com/biron/singerTargetClickhouse/JsonSchemaInspector.kt
package com.biron.singerTargetClickhouse

import com.biron.singer.core.domain.JsonSchema
import com.biron.singer.core.domain.typeNullable
import io.github.oshai.kotlinlogging.KotlinLogging
import java.security.MessageDigest

private val logger = KotlinLogging.logger {}

const val NESTED_SUB_OBJECT_SEPARATOR = "\$%€£"

enum class PKType { ROOT, PARENT, CURRENT, LEVEL }

typealias ValueExtractor = (Any?) -> Any?

data class ColumnMap(
	val prop: String?,
	val valueExtractor: ValueExtractor,
	val sqlIdentifier: String,
	val chType: String?,
	val schemaType: String?,
	val typeFormat: String?,
	val nullable: Boolean,
	val lowCardinality: Boolean,
	val nestedArray: Boolean,
)

data class PkMap(
	val prop: String,
	val valueExtractor: ValueExtractor,
	val sqlIdentifier: String,
	val chType: String?,
	val schemaType: String?,
	val typeFormat: String?,
	val nullable: Boolean,
	val lowCardinality: Boolean,
	val nestedArray: Boolean,
	val pkType: PKType,
)

data class SourceMeta(
	val prop: String,
	val children: List<SourceMeta>,
	val pkMappings: List<PkMap>,
	val simpleColumnMappings: List<ColumnMap>,
	val sqlTableName: String,
	val cleaningColumn: String? = null,
)

data class JsonSchemaInspectorContext(
	val alias: String,
	val schema: JsonSchema,
	val keyProperties: List<String>,
	val subtableSeparator: String = "__",
	val parentCtx: JsonSchemaInspectorContext? = null,
	val level: Int = 0,
	val tableName: String = defaultTableName(alias, subtableSeparator, parentCtx),
	val cleaningColumn: String? = null,
	val allKeyProperties: SchemaKeyProperties = SchemaKeyProperties.empty,
) {
	fun isRoot(): Boolean = parentCtx == null
	val rootCtx: JsonSchemaInspectorContext by lazy { generateSequence(this) { it.parentCtx }.last() }

	companion object {
		fun defaultTableName(alias: String, subtableSeparator: String, parentCtx: JsonSchemaInspectorContext?): String =
			"${parentCtx?.let { "${it.tableName}$subtableSeparator" }.orEmpty()}$alias"
	}
}

private fun formatLevelIndexColumn(level: Int): String = "_level_${level}_index"
fun formatRootPKColumn(prop: String): String = "_root_$prop"
private fun formatParentPKColumn(prop: String): String = "_parent_$prop"

fun escapeIdentifier(id: String, subtableSeparator: String = "__"): String =
	id.replace(NESTED_SUB_OBJECT_SEPARATOR, subtableSeparator)
		.let {
			if (it.length > 64) {
				val uid = sha1Hex(it).substring(0, 10)
				it.substring(0, 64 - uid.length - 27) + uid + it.substring(it.length - 27)
			} else it
		}
		.let { "`$it`" }

private fun sha1Hex(input: String): String =
	MessageDigest.getInstance("SHA-1").digest(input.toByteArray())
		.joinToString("") { "%02x".format(it) }

fun buildMeta(ctx: JsonSchemaInspectorContext): SourceMeta {
	val (simpleColumnMappings, children) = buildMetaProps(ctx)
	return SourceMeta(
		prop = ctx.alias,
		sqlTableName = escapeIdentifier(ctx.tableName, ctx.subtableSeparator),
		pkMappings = buildMetaPkProps(ctx),
		cleaningColumn = ctx.cleaningColumn,
		simpleColumnMappings = simpleColumnMappings,
		children = children,
	)
}

private data class MetaProps(
	val simpleColumnMappings: List<ColumnMap>,
	val children: List<SourceMeta>,
)

private fun buildMetaPkProps(ctx: JsonSchemaInspectorContext): List<PkMap> = buildList {
	// _root_X (only on non-root nodes)
	ctx.rootCtx.takeIf { it != ctx }?.keyProperties?.forEach { prop ->
		add(buildMetaPkProp(prop, ctx.rootCtx, PKType.ROOT, ::formatRootPKColumn))
	}
	// _parent_X (if parent has all_key_properties non-empty)
	ctx.parentCtx?.keyProperties?.takeIf { ctx.parentCtx.allKeyProperties.props.isNotEmpty() }?.forEach { prop ->
		add(buildMetaPkProp(prop, ctx.parentCtx, PKType.PARENT, ::formatParentPKColumn))
	}
	// Current-level key properties
	ctx.keyProperties.forEach { prop ->
		add(buildMetaPkProp(prop, ctx, PKType.CURRENT))
	}
	// level_N_index
	repeat(ctx.level) { level ->
		val prop = formatLevelIndexColumn(level)
		add(
			PkMap(
				prop = prop,
				valueExtractor = buildValueExtractor(prop),
				sqlIdentifier = escapeIdentifier(prop, ctx.subtableSeparator),
				chType = "Int32",
				schemaType = null,
				typeFormat = null,
				nullable = false,
				lowCardinality = false,
				nestedArray = false,
				pkType = PKType.LEVEL,
			)
		)
	}
}

private fun buildMetaPkProp(
	prop: String,
	ctx: JsonSchemaInspectorContext,
	pkType: PKType,
	fieldFormatter: ((String) -> String)? = null,
): PkMap {
	val colType = getSimpleColumnType(ctx, prop)
	return PkMap(
		prop = prop,
		valueExtractor = buildValueExtractor(prop),
		sqlIdentifier = escapeIdentifier(fieldFormatter?.invoke(prop) ?: prop, ctx.subtableSeparator),
		chType = colType?.chType,
		schemaType = colType?.schemaType,
		typeFormat = colType?.typeFormat,
		nullable = false,
		lowCardinality = false,
		nestedArray = false,
		pkType = pkType,
	)
}

private fun buildValueExtractor(prop: String?): ValueExtractor {
	if (prop == null) return { data -> data }
	val parts = prop.split(NESTED_SUB_OBJECT_SEPARATOR)
	if (parts.size == 1) {
		val onlyPart = parts[0]
		return { data -> (data as? Map<*, *>)?.get(onlyPart) }
	}
	return { data ->
		parts.fold(data) { acc, part -> (acc as? Map<*, *>)?.get(part) }
	}
}

private fun buildMetaProps(ctx: JsonSchemaInspectorContext): MetaProps =
	if ("object" in ctx.schema.type) buildObjectMetaProps(ctx)
	else buildScalarMetaProps(ctx)

private fun buildObjectMetaProps(ctx: JsonSchemaInspectorContext): MetaProps =
	(ctx.schema.properties ?: emptyMap())
		.filterKeys { it !in ctx.keyProperties }
		.entries
		.fold(MetaProps(emptyList(), emptyList())) { acc, (key, propDef) ->
			val propDefTypes = propDef.type
			when {
				"object" in propDefTypes -> {
					val flattened = flattenNestedObject(propDef, key, ctx)
					acc.copy(
						simpleColumnMappings = acc.simpleColumnMappings + flattened.simpleColumnMappings,
						children = acc.children + flattened.children,
					)
				}

				"array" in propDefTypes && propDef.format != "nested" -> {
					if (ctx.rootCtx.keyProperties.isEmpty() && ctx.rootCtx.allKeyProperties.props.isEmpty())
						throwError(ctx, "$key refused: array child with no root key properties")
					acc.copy(children = acc.children + createSubTable(propDef, key, ctx))
				}

				else -> {
					val colType = getSimpleColumnType(ctx, key)
					if (colType != null) {
						acc.copy(
							simpleColumnMappings = acc.simpleColumnMappings + ColumnMap(
								prop = key,
								valueExtractor = buildValueExtractor(key),
								sqlIdentifier = escapeIdentifier(key, ctx.subtableSeparator),
								chType = colType.chType,
								schemaType = colType.schemaType,
								typeFormat = colType.typeFormat,
								nullable = colType.nullable,
								lowCardinality = colType.lowCardinality,
								nestedArray = colType.nestedArray,
							),
						)
					} else {
						logger.warn { "'${ctx.alias}': '$key': could not be registered (type '${propDef.type}' unrecognized)" }
						acc
					}
				}
			}
		}

private fun buildScalarMetaProps(ctx: JsonSchemaInspectorContext): MetaProps {
	if (ctx.schema.type.isEmpty()) return MetaProps(emptyList(), emptyList())
	val colType = getSimpleColumnType(ctx, null) ?: return MetaProps(emptyList(), emptyList())
	return MetaProps(
		simpleColumnMappings = listOf(
			ColumnMap(
				prop = null,
				valueExtractor = buildValueExtractor(null),
				sqlIdentifier = escapeIdentifier("value", ctx.subtableSeparator),
				chType = colType.chType,
				schemaType = colType.schemaType,
				typeFormat = colType.typeFormat,
				nullable = ctx.schema.typeNullable,
				lowCardinality = false,
				nestedArray = false,
			),
		),
		children = emptyList(),
	)
}

private fun flattenNestedObject(propDef: JsonSchema, key: String, ctx: JsonSchemaInspectorContext): MetaProps {
	val nullable = propDef.typeNullable
	val mergedProperties: Map<String, JsonSchema> = (propDef.properties ?: emptyMap())
		.map { (nestedKey, nested) ->
			"$key$NESTED_SUB_OBJECT_SEPARATOR$nestedKey" to
					nested.copy(type = if (nullable) makeNullable(nested.type) else nested.type)
		}
		.toMap()

	val nestedSchema = JsonSchema(type = listOf("object"), properties = mergedProperties)
	return buildMetaProps(
		JsonSchemaInspectorContext(
			alias = ctx.alias,
			schema = nestedSchema,
			keyProperties = emptyList(),
			subtableSeparator = ctx.subtableSeparator,
			parentCtx = ctx,
			level = ctx.level,
			tableName = ctx.tableName,
		),
	)
}

private fun createSubTable(propDef: JsonSchema, key: String, ctx: JsonSchemaInspectorContext): SourceMeta {
	val itemSchema = propDef.items ?: JsonSchema(type = listOf("string"))
	val childAllKeyProps = ctx.allKeyProperties.children[key]
	return buildMeta(
		JsonSchemaInspectorContext(
			alias = key,
			schema = itemSchema,
			keyProperties = childAllKeyProps?.props ?: emptyList(),
			subtableSeparator = ctx.subtableSeparator,
			parentCtx = ctx,
			level = ctx.level + 1,
			allKeyProperties = childAllKeyProps ?: SchemaKeyProperties.empty,
		),
	)
}

private data class SimpleColumnType(
	val chType: String?,
	val schemaType: String?,
	val typeFormat: String?,
	val nullable: Boolean,
	val lowCardinality: Boolean,
	val nestedArray: Boolean,
)

private fun getSimpleColumnType(ctx: JsonSchemaInspectorContext, key: String?): SimpleColumnType? {
	var propDef = if (key != null) ctx.schema.properties?.get(key) else ctx.schema
	if (propDef == null) {
		throwError(ctx, "Key '$key' does not match any usable prop in schema props '${ctx.schema.properties}'")
	}
	var nestedArray = false
	if (propDef.format == "nested" && "array" in propDef.type) {
		propDef = propDef.items ?: return null
		nestedArray = true
	}
	val type = propDef.type.firstOrNull { it != "null" }
	val chType = getSimpleColumnSqlType(ctx, propDef, key) ?: return null
	return SimpleColumnType(
		chType = chType,
		schemaType = type,
		typeFormat = propDef.format,
		nullable = propDef.typeNullable,
		lowCardinality = propDef.lowCardinality == true,
		nestedArray = nestedArray,
	)
}

fun getSimpleColumnSqlType(ctx: JsonSchemaInspectorContext, propDef: JsonSchema, key: String? = null): String? {
	val type = propDef.type.firstOrNull { it != "null" }
	val format = propDef.format
	return when (type) {
		"string" -> when (format) {
			"date", "x-excel-date" -> "Date"
			"date-time" -> "DateTime"
			"date-time64" -> "DateTime64"
			"uuid" -> "UUID"
			else -> "String"
		}

		"integer" -> when (format) {
			null -> "Int64"
			"int128" -> "Int128"
			"int64" -> "Int64"
			"int32" -> "Int32"
			"int16" -> "Int16"
			"int8" -> "Int8"
			else -> throwError(ctx, "$key: unsupported integer format [$format]")
		}

		"number" -> when (format) {
			null -> "Decimal(${propDef.precision ?: 16}, ${propDef.decimals ?: 2})"
			"float64" -> "Float64"
			"float32" -> "Float32"
			else -> throwError(ctx, "$key: unsupported number format [$format]")
		}

		"boolean" -> when (format) {
			null -> "UInt8"
			else -> throwError(ctx, "$key: unsupported number format [$format]")
		}

		else -> null
	}
}

private fun makeNullable(types: List<String>): List<String> = when {
	types.isEmpty() -> emptyList()
	"null" in types -> types
	else -> types + "null"
}

private fun throwError(ctx: JsonSchemaInspectorContext, msg: String, childAlias: String? = null): Nothing {
	var current = ctx
	var alias = "${ctx.alias}${childAlias?.let { ".$it" } ?: ""}"
	while (current.parentCtx != null) {
		current = current.parentCtx!!
		alias = "${current.alias}.$alias"
	}
	logger.error { "$alias: $msg" }
	error("$alias: $msg")
}
